synapse/handlers/room.py

```diff
@@ -1635,6 +1606,20 @@ async def get_event_for_timestamp(
                 errcode=Codes.NOT_FOUND,
             )
 
+        if local_event.internal_metadata.outlier:
+            # We might already have these on hand from asking a remote server before
+            if likely_domains is None:
+                likely_domains = await self._storage_controllers.state.get_current_hosts_in_room_ordered(
+                    room_id
+                )
+
+            # Backfill this event so we can get a pagination token for
+            # it with `/context` and paginate `/messages` from this
+            # point.
+            await self.federation_event_handler.backfill_event_id(
+                likely_domains, room_id, local_event.event_id
+            )
```

Comment:
The meat of the fix is here. The rest is refactoring to make this easy and can be reviewed commit-by-commit

Reply:
TODO: Remove this. We don't actually want to work with `outliers` at all here, see https://github.com/matrix-org/synapse/pull/14096#issuecomment-1281561270

Reply:
Removed ✅